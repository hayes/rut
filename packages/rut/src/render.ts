import React from 'react';
import { act } from 'react-test-renderer';
import Renderer from './Renderer';
import { UnknownProps, RendererOptions } from './types';

// function render<Comp extends React.Component, Props = Comp['props']>(
//   element: React.ReactElement<Props>,
// ): RutRenderer<Props>;
function render<Props = UnknownProps>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Renderer<Props> {
  let renderer: Renderer<Props>;

  act(() => {
    renderer = new Renderer(element, options);
  });

  return renderer!;
}

export default render;
