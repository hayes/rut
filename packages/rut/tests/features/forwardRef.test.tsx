import React from 'react';
import render from '../../src/render';

describe('ForwardRef', () => {
  const mock = { tagName: 'BUTTON' };

  describe('common', () => {
    const Button = React.forwardRef(
      ({ children }: { children: React.ReactNode }, ref: React.Ref<HTMLButtonElement>) => {
        return (
          <button type="button" ref={ref}>
            {children}
          </button>
        );
      },
    );

    it('resolves the ref from the inner component', () => {
      const ref = React.createRef<HTMLButtonElement>();

      expect(ref.current).toBeNull();

      const wrapper = render(<Button ref={ref}>Child</Button>, {
        refs: { button: mock },
      });

      expect(wrapper.debug()).toMatchSnapshot();
      expect(ref.current).toBe(mock);
    });
  });

  describe('HOC', () => {
    function withRef(WrappedComponent: React.ComponentType<{ children: React.ReactNode }>): any {
      class WithRef extends React.Component<{ forwardedRef?: React.Ref<unknown> | null }> {
        render() {
          const { forwardedRef, ...rest } = this.props;

          return <WrappedComponent ref={forwardedRef} {...rest} />;
        }
      }

      return React.forwardRef((props, ref) => {
        return <WithRef {...props} forwardedRef={ref} />;
      });
    }

    class InnerButton extends React.Component<{ children: React.ReactNode }> {
      render() {
        return <button type="button">{this.props.children}</button>;
      }
    }

    const Button = withRef(InnerButton);

    it('resolves the ref from the HOC', () => {
      const ref = React.createRef<HTMLButtonElement>();

      expect(ref.current).toBeNull();

      const wrapper = render(<Button ref={ref}>Child</Button>);

      expect(wrapper.debug()).toMatchSnapshot();
      expect(ref.current!.constructor.name).toBe('InnerButton');
    });
  });
});
