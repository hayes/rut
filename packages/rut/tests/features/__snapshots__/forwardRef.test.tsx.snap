// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForwardRef HOC resolves the ref from the HOC 1`] = `
"<WithRef forwardedRef={(InnerButton)}>
  <InnerButton>
    <button type=\\"button\\">
      Child
    </button>
  </InnerButton>
</WithRef>"
`;

exports[`ForwardRef common resolves the ref from the inner component 1`] = `
"<button type=\\"button\\">
  Child
</button>"
`;
